src/snowflake/snowpark/_internal/temp_table_auto_cleaner.py

```diff
@@ -54,53 +46,26 @@ def _delete_ref_count(self, name: str) -> None:
         self.ref_count_map[name] -= 1
         if self.ref_count_map[name] == 0:
             self.ref_count_map.pop(name)
-            # clean up
-            self.queue.put(name)
+            if self.session.auto_clean_up_temp_table_enabled:
+                self.drop_table(name)
         elif self.ref_count_map[name] < 0:
             logging.debug(
                 f"Unexpected reference count {self.ref_count_map[name]} for table {name}"
             )
 
-    def process_cleanup(self) -> None:
-        while not self.stop_event.is_set():
-            try:
-                # it's non-blocking after timeout and become interruptable with stop_event
-                # it will raise an `Empty` exception if queue is empty after timeout,
-                # then we catch this exception and avoid breaking loop
-                table_name = self.queue.get(timeout=1)
-                self.drop_table(table_name)
-            except Empty:
-                continue
-
     def drop_table(self, name: str) -> None:
         common_log_text = f"temp table {name} in session {self.session.session_id}"
-        logging.debug(f"Cleanup Thread: Ready to drop {common_log_text}")
+        logging.debug(f"Ready to drop {common_log_text}")
         try:
-            # TODO SNOW-1556553: Remove this workaround once multi-threading of Snowpark session is supported
-            with self.session._conn._conn.cursor() as cursor:
-                cursor.execute(
-                    f"drop table if exists {name} /* internal query to drop unused temp table */",
-                    _statement_params={DROP_TABLE_STATEMENT_PARAM_NAME: name},
-                )
-            logging.debug(f"Cleanup Thread: Successfully dropped {common_log_text}")
+            async_job = self.session.sql(
+                f"drop table if exists {name} /* internal query to drop unused temp table */",
+            )._internal_collect_with_tag_no_telemetry(
+                block=False, statement_params={DROP_TABLE_STATEMENT_PARAM_NAME: name}
+            )
+            logging.debug(
+                f"Dropping {common_log_text} with query id {async_job.query_id}"
+            )
         except Exception as ex:
             logging.warning(
-                f"Cleanup Thread: Failed to drop {common_log_text}, exception: {ex}"
+                f"Failed to drop {common_log_text}, exception: {ex}"
```

Comment:
we can probably recored a telemetry later for failed table drops

Reply:
actually we might not need telemetry for this case, because we can group by temp table name from statement params, then we can find the corresponding query is successful or not.

Reply:
you can only do this for queries that has reached to server, not for the ones that don't even have a query id.
i think what we can do here is check if the async job is not None, and if it doesn't have a query_id, if not, we can sent a separate telemetry over

Reply:
async job will not be None and only query id might be None. I added such a telemetry.